R/editing.R

```diff
@@ -813,3 +813,154 @@ set_file_name <- function(mn, pid, name) {
+#' @param new_data_path (character) Path to new data object.
+#' @param resource_map_pid (character) PID for resource map to update.
+#' @param format_id (character) Optional. The format ID to set for the object.
+#' When not set, \code{\link{guess_format_id}} will be used
+#' to guess the format ID. Should be a \href{https://cn.dataone.org/cn/v2/formats}{DataONE format ID}.
+#' @param public (logical) Optional. Make the update public. If FALSE,
+#' will set the metadata and resource map to private (but not the data objects).
+#' This applies to the new metadata PID and its resource map and data object.
+#' Access policies are not affected.
+#' @param use_doi (logical) Optional. If TRUE, a new DOI will be minted.
+#' @param ... Other arguments to pass into \code{\link{publish_update}}.
+#'
+#' @return PIDs (character) Named character vector of PIDs in the data package, including PIDs
+#' for the metadata, resource map, and data objects.
+#'
+#' @keywords update_object publish_update
+#'
+#' @import dataone
+#' @import EML
+#'
+#' @examples
+#' \dontrun{
+#' cnTest <- dataone::CNode("STAGING")
+#' mnTest <- dataone::getMNode(cnTest,"urn:node:mnTestARCTIC")
+#'
+#' pkg <- create_dummy_package_full(mnTest, title = "My package")
+#'
+#' file.create("new_file.csv")
+#' update_package_object(mnTest, pkg$data[1], "new_file.csv", pkg$resource_map, format_id = "text/csv")
+#' file.remove("new_file.csv")
+#' }
+#'
+#' @export
+update_package_object <- function(mn,
+                                  data_pid,
+                                  new_data_path,
+                                  resource_map_pid,
+                                  format_id = NULL,
+                                  public = TRUE,
+                                  use_doi = FALSE,
+                                  ...) {
+  stopifnot(is(mn, "MNode"))
+  stopifnot(is.character(data_pid), nchar(data_pid) > 0)
+  stopifnot(is.character(new_data_path), nchar(new_data_path) > 0, file.exists(new_data_path))
+  stopifnot(is.character(resource_map_pid), nchar(resource_map_pid) > 0)
+  stopifnot(is.logical(public))
+
+  pkg <- get_package(mn, resource_map_pid)
+  eml <- EML::read_eml(rawToChar(dataone::getObject(mn, pkg$metadata)))
+
+  new_data_pid <- update_object(mn,
+                                pid = data_pid,
+                                path = new_data_path,
+                                format_id = format_id)
+
+  other_data_pids <- pkg$data[which(pkg$data != data_pid)] # wrapped in which for better NA handling
+  new_data_pids <- c(other_data_pids, new_data_pid)
+
+  eml_new <- tryCatch(update_physical(eml = eml,
+                                      mn = mn,
+                                      data_pid = data_pid,
+                                      new_data_pid = new_data_pid),
+                      error = function(e) {
+                        print("The data PID does not match any physical sections, so the EML will not be updated.")
```

Comment:
a little nitpicky but I feel like the error message here could be more clear - maybe "The obsoleted data PID..."